### DISCUSSION TITLE
How to join with custom condition using QueryBuilder

### DISCUSSION THREAD

**Discussion by emulienfou** [mikro-orm/mikro-orm#4913] [Q&A] [answered]
Hello there, I'm currently searching for help regarding a join  with custom condition.

QueryBuilder is proving multiple functions that take in 3rd parameter a join condition like: `cond?: QBFilterQuery` (e.g. leftJoin).
Mikro-Orm documentation doesn't show any example on how to use this 3rd parameter to add custom/complex condition.

Let's take an example:
> I want to be able to join table **Performance** with a custom condition including an **OR**.

What I'm trying to achieve is something in SQL:
```sql
SELECT "Edition".*
FROM "Edition"
LEFT JOIN "Stage" ON "Edition"."id" = "Stage"."editionId"
LEFT JOIN "Performance" ON "Performance"."stageId" = "Stage"."id" OR "Performance"."editionId" = "Edition"."id";
```

Using the QueryBuilder I would do something like:
```ts
em.createQueryBuilder("Edition")
     .leftJoin("Edition.stages", "Stage")
     .leftJoin("Stages.performances", "Performance", {
        $or: [
          { "Performance.stageId": { $eq: "Stage.id" } },
          { "Performance.editionId": { $eq: "Edition.id" } },
        ],
      })
```

However this code is returning the next error:
> **invalid input syntax for type uuid: "Stage.id"**

And seems to generate a LEFT JOIN query like:
> ** left join "Performance" on "Stage"."id" = "Performance"."stageId" and ("Performance"."stageId" = 'Stage.id' or "Performance"."editionId" = 'Edition.id')**

Which contains both **AND** and **OR** condition instead of just the custom condition I'm writing.

**Questions are:**
---
* How can I , by using the query builder add a custom condition like demonstrated in the example above?
* How do I tell the query builder to use my custom condition without the default join ON part: `"Stage"."id" = "Performance"."stageId" and` ?

**Comment by B4nan** [comment]
Have you seen the docs?

https://mikro-orm.io/docs/query-builder

**Reply by emulienfou** [reply]
Yes, the documentation for QueryBuilder is only referencing to join/leftJoin/leftJoinAndSelect with 2 parameters only.
I was not able to find any examples in docs or Github issues/code demonstrating how the `cond` 3rd param is working actually!

Or did I didn't see it!?!

**Reply by B4nan** [reply]
https://mikro-orm.io/docs/query-builder#referring-to-column-in-update-queries

**Reply by emulienfou** [reply]
Thanks for the link for this section of the docs.
This seems to fix my question 1, but I'm still having the default value with the Join query.

I replaced the leftJoin code by:
```ts
.leftJoin("Stages.performances", "Performance",
  qb1.raw(`"Performance"."stageId" = "Stage"."id" OR "Performance"."editionId" = "Edition"."id"`)
)
```

However this is generating something like:
```sql
left join "Performance" on "Stage"."id" = "Performance"."stageId" and "Edition"."id" = "Performance"."stageId" = "s"."id" OR "Performance"."editionId" = "Edition"."id"
```
it seems to generate automatically this part of the query: `"Stage"."id" = "Performance"."stageId" and "Edition"."id" =`

**Reply by B4nan** [reply]
you will always have the FKs in the join conditions, that is not configurable. the QB is designed to work with your entities, not with random queries.

**Reply by B4nan** [reply, marked as answer]
joining a raw query is supported only in v6

https://mikro-orm.io/docs/next/query-builder#joining-sub-queries

check the QB tests for more examples to see whats possible

https://github.com/mikro-orm/mikro-orm/blob/master/tests/QueryBuilder.test.ts#L2024

as you can see in the first test, if you use a subquery join, it wont automatically add the default join condition.

**Reply by emulienfou** [reply]
Thanks for your time and answers @B4nan , reading more the documentation on Mikro-Orm website and links your shared help me found that I was able to run my query by writing something like:
```ts
const qb = em.createQueryBuilder("Edition");

qb.leftJoin("Edition.stages", "Stage")
  .leftJoin("Stages.performances", "Performance", {
        $or: [
          { "Performances.stageId": { $eq: qb.ref("Stage.id") } },
          { "Performances.editionId": { $eq: qb.ref("Edition.id") } },
        ],
      })
```

Sure like you said, the `FKs in the join conditions` will not be removed, but this code seems to do the job correctly.
